### PR TITLE
Function overloading.

### PR DESCRIPTION
**Pull request by chriseth**
Finishes work on #1162.

### CODE REVIEW

**Comment by gavofyork** [comment]
rebase?

**Comment by chriseth** [comment]
Used merge instead of rebase as rebasing is too complicated given the history of this pull request.

File: libsolidity/NameAndTypeResolver.cpp [hunk 1]
```diff
@@ -90,31 +115,51 @@ void NameAndTypeResolver::updateDeclaration(Declaration const& _declaration)
 	solAssert(_declaration.getScope() == nullptr, "Updated declaration outside global scope.");
 }
 
-Declaration const* NameAndTypeResolver::resolveName(ASTString const& _name, Declaration const* _scope) const
+set<Declaration const*> NameAndTypeResolver::resolveName(ASTString const& _name, Declaration const* _scope) const
 {
 	auto iterator = m_scopes.find(_scope);
 	if (iterator == end(m_scopes))
-		return nullptr;
+		return std::set<Declaration const*>({});
```
Comment (LianaHus) [inline libsolidity/NameAndTypeResolver.cpp]: std:: can be removed

File: libsolidity/NameAndTypeResolver.cpp [hunk 2]
```diff
@@ -361,24 +413,31 @@ bool ReferencesResolver::visit(Mapping&)
...
-	Declaration const* declaration = m_resolver.getNameFromCurrentScope(_typeName.getName());
-	if (!declaration)
+	auto declarations = m_resolver.getNameFromCurrentScope(_typeName.getName());
+	if (declarations.empty())
 		BOOST_THROW_EXCEPTION(DeclarationError() << errinfo_sourceLocation(_typeName.getLocation())
 												 << errinfo_comment("Undeclared identifier."));
-	_typeName.setReferencedDeclaration(*declaration);
+	else if (declarations.size() > 1)
+		BOOST_THROW_EXCEPTION(DeclarationError() << errinfo_sourceLocation(_typeName.getLocation())
+												 << errinfo_comment("Duplicate identifier."));
```
Comment (LianaHus) [inline libsolidity/NameAndTypeResolver.cpp]: indent

File: libsolidity/DeclarationContainer.cpp [hunk 3]
```diff
@@ -34,17 +35,34 @@ bool DeclarationContainer::registerDeclaration(Declaration const& _declaration,
...
+		m_invisibleDeclarations[name].insert(&_declaration);
 	else
-		m_declarations[name] = &_declaration;
+		m_declarations[name].insert(&_declaration);
+
 	return true;
 }
 
-Declaration const* DeclarationContainer::resolveName(ASTString const& _name, bool _recursive) const
+std::set<Declaration const*> DeclarationContainer::resolveName(ASTString const& _name, bool _recursive) const
```
Comment (LianaHus) [inline libsolidity/DeclarationContainer.cpp]: here to (std::)

File: libsolidity/DeclarationContainer.cpp [hunk 4]
```diff
@@ -52,7 +70,7 @@ Declaration const* DeclarationContainer::resolveName(ASTString const& _name, boo
 		return result->second;
 	if (_recursive && m_enclosingContainer)
 		return m_enclosingContainer->resolveName(_name, true);
-	return nullptr;
+	return std::set<Declaration const*>({});
```
Comment (LianaHus) [inline libsolidity/DeclarationContainer.cpp]: std::

**Comment by chriseth** [comment]
No failures for the solidity tests.

**Comment by chriseth** [comment]
Merging since all issues addressed.

**Pull request merged**